src/rt/alloc.rs

```diff
@@ -0,0 +1,64 @@
+use crate::rt;
+use crate::rt::object::Object;
+
+/// Tracks an allocation
+#[derive(Debug)]
+pub(crate) struct Allocation {
+    obj: Object,
+}
+
+#[derive(Debug)]
+pub(super) struct State {
+    is_dropped: bool,
+}
+
+/// Track a raw allocation
+pub(crate) fn alloc(ptr: *mut u8) {
+    rt::execution(|execution| {
+        let obj = execution.objects.insert_alloc(State { is_dropped: false });
+
+        let allocation = Allocation { obj };
+
+        let prev = execution.raw_allocations.insert(ptr as usize, allocation);
+        assert!(prev.is_none(), "pointer already tracked");
+    });
+}
+
+/// Track a raw deallocation
+pub(crate) fn dealloc(ptr: *mut u8) {
```

Comment:
Here's a thought (not a blocker for this PR): it would be really cool if `loom`'s leak-detection code could eventually capture the backtrace where an object was allocated from and display it to the user if that allocation was leaked. Could be worth doing in a follow-up...